crates/nu-cli/src/completions/completer.rs

```diff
@@ -378,36 +380,53 @@ impl NuCompleter {
                                     let arg_pos = positional_arg_indices.len();
                                     signature
                                         .get_positional(arg_pos)
-                                        .and_then(|pos_arg| pos_arg.custom_completion)
+                                        .and_then(|pos_arg| pos_arg.custom_completion.clone())
                                 }
                                 _ => None,
                             }
                         };
 
-                        if let Some(decl_id) = custom_completion_decl_id {
-                            // for `--foo <tab>` and `--foo=<tab>`, the arg span should be trimmed
-                            let (new_span, prefix) = if matches!(arg, Argument::Named(_)) {
-                                strip_placeholder_with_rsplit(
-                                    working_set,
-                                    &span,
-                                    |b| *b == b'=' || *b == b' ',
-                                    strip,
-                                )
-                            } else {
-                                strip_placeholder_if_any(working_set, &span, strip)
-                            };
-                            let ctx = Context::new(working_set, new_span, prefix, offset);
-
-                            let mut completer = CustomCompletion::new(
-                                decl_id,
-                                prefix_str.into(),
-                                pos - offset,
-                                FileCompletion,
-                            );
-
-                            // Prioritize argument completions over (sub)commands
-                            suggestions.splice(0..0, self.process_completion(&mut completer, &ctx));
-                            break;
+                        match custom_completion {
+                            Some(Completion::Command(decl_id)) => {
+                                let (new_span, prefix) = if matches!(arg, Argument::Named(_)) {
+                                    strip_placeholder_with_rsplit(
+                                        working_set,
+                                        &span,
+                                        |b| *b == b'=' || *b == b' ',
+                                        strip,
+                                    )
+                                } else {
+                                    strip_placeholder_if_any(working_set, &span, strip)
+                                };
+                                let ctx = Context::new(working_set, new_span, prefix, offset);
```

Comment:
How about move the shared snippets into another if/match block and keep the comments?

Especially the `strip_placeholder_with_rsplit` stuff is not very intuitive in my opinion